pywhy_stats/discrepancy/base.py

```diff
@@ -0,0 +1,127 @@
+from typing import Callable, Optional
+
+import numpy as np
+from joblib import Parallel, delayed
+from numpy.typing import ArrayLike
+from sklearn.base import BaseEstimator
+from sklearn.linear_model import LogisticRegression
+
+from pywhy_stats.kernel_utils import _default_regularization
+
+
+def _preprocess_propensity_data(
+    group_ind: ArrayLike,
+    propensity_model: Optional[BaseEstimator],
+    propensity_weights: Optional[ArrayLike],
+):
+    if group_ind.ndim != 1:
+        raise RuntimeError("group_ind must be a 1d array.")
+    if len(np.unique(group_ind)) != 2:
+        raise RuntimeError(
+            f"There should only be two groups. Found {len(np.unique(group_ind))} groups."
+        )
+    if propensity_model is not None and propensity_weights is not None:
+        raise ValueError(
+            "Both propensity model and propensity estimates are specified. Only one is allowed."
+        )
+    if propensity_weights is not None:
+        if propensity_weights.shape[0] != len(group_ind):
+            raise ValueError(
+                f"There are {propensity_weights.shape[0]} pre-defined estimates, while "
+                f"there are {len(group_ind)} samples."
+            )
+        if propensity_weights.shape[1] != len(np.unique(group_ind.squeeze())):
+            raise ValueError(
+                f"There are {propensity_weights.shape[1]} group pre-defined estimates, while "
+                f"there are {len(np.unique(group_ind))} unique groups."
+            )
+
+
+def _compute_propensity_scores(
+    group_ind: ArrayLike,
+    propensity_model: Optional[BaseEstimator] = None,
+    propensity_weights: Optional[ArrayLike] = None,
+    n_jobs: Optional[int] = None,
+    random_state: Optional[int] = None,
+    **kwargs,
+):
+    if propensity_model is None:
+        K: ArrayLike = kwargs.get("K")
+
+        # compute a default penalty term if using a kernel matrix
+        if K.shape[0] == K.shape[1]:
+            propensity_penalty_ = _default_regularization(K)
+            C = 1 / (2 * propensity_penalty_)
```

Comment:
This just allows us to magically infer a default regularization for LogisticRegression for propensity score estimation. Since LogisticRegression is sort of the de-facto default in literature, this seems like a reasonable default. Alternatively, `propensity_model` can be specified by the user fully, which can be for example a fully specified RandomForestClassifier.

Reply:
I added some inline comments too